File: example/src/sigmoid.py
from .tensor import tensor

class sigmoid:
  def __repr__(self):
    return 'sigmoid()'

  def forward(self, input):
    return 1 / (1 + tensor.exp(-input))
    
  def backward(self, input):
    return tensor.exp(-input) / ((1 + tensor.exp(-input)) ** 2)

File: efficiency-testing/aiinpycpp/sigmoid.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

class sigmoid {
  public:
    vector <double> forward(vector <double> input) {
      transform(input.begin(), input.end(), input.begin(), [](double input){ return 1 / (1 + exp(-input)); });
      return input;
    }

    vector <double> backward(vector <double> input) {
      transform(input.begin(), input.end(), input.begin(), [](double input){ return input * (1 - input); });
      return input;
    }
};
